src/App.tsx

```diff
@@ -4,32 +4,82 @@ import { SubmitLink } from "./components/SubmitLink";
 import { Message } from "./components/Message";
 import { Listing } from "./components/Listing";
 import { ListItem } from "./types/ListItem";
-import "./App.css";
+import "./App";
+import { Tag } from "./types/Tag";
+import { COLORS } from "./utils/constants";
 
 const App = (): JSX.Element => {
   const submitForm = () => {
     const currentDate = new Date();
+    const newItems = [...items];
     setItems([
-      ...items,
+      ...newItems,
       {
-        id: (items.length + 1).toString(),
+        id: (newItems.length + 1).toString(),
         url: url,
         date: currentDate.toLocaleString(),
         alreadyRead: false,
+        tags: [],
       },
     ]);
     setUrl("");
   };
   const changeItemStatus = (listItem: ListItem) => {
-    items.splice(
-      items.findIndex((it) => it.id === listItem.id),
+    const newItems = [...items];
+    const newListItem = { ...listItem, alreadyRead: !listItem.alreadyRead };
+    newItems.splice(
+      newItems.findIndex((it) => it.id === newListItem.id),
       1,
-      { ...listItem, alreadyRead: !listItem.alreadyRead }
+      newListItem
     );
-    setItems([...items]);
+    setItems([...newItems]);
+  };
+  const updateKnownTags = (title: string) => {
+    const tagFound = knownTags.find((t) => t.title === title);
+    let newTag = undefined;
+    if (!tagFound) {
+      const tagId = knownTags.length + 1;
+      const tagColor =
+        COLORS[tagId >= COLORS.length ? COLORS.length - 1 : tagId - 1];
+      newTag = {
+        id: tagId.toString(),
+        title: title,
+        color: tagColor,
+      };
+      setKnownTags([...knownTags, newTag]);
+    }
+    return newTag;
+  };
+  const addTag = (listItem: ListItem, tag: Tag) => {
+    const newItems = [...items];
+    const newListItem = {
+      ...listItem,
+      tags: [...listItem.tags, tag],
+    };
+    newItems.splice(
+      newItems.findIndex((it) => it.id === newListItem.id),
+      1,
+      newListItem
+    );
+    setItems([...newItems]);
+  };
+  const deleteTag = (listItem: ListItem, tagItem: Tag) => {
+    const newItems = [...items];
+    const newListItem = { ...listItem, tags: [...listItem.tags] };
+    newListItem.tags.splice(
+      newListItem.tags.findIndex((li) => li.id === tagItem.id),
+      1
+    );
+    newItems.splice(
+      newItems.findIndex((it) => it.id === newListItem.id),
+      1,
+      newListItem
+    );
+    setItems([...newItems]);
   };
   const [url, setUrl] = useState("");
   const [items, setItems] = useState<ListItem[]>([]);
+  const [knownTags, setKnownTags] = useState<Tag[]>([]);
   return (
     <div>
       <AddLink url={url.toLowerCase()} onChange={setUrl} />
```

Comment:
Not part of this PR but I believe that the `toLowerCase()` call should hapen within `<AddLink .../>` instead of being called like this. Not a big deal but I think it's a bit cleaner that way.

Reply:
Right!